protocol-tokio/src/protocol/codec/message.rs

```diff
@@ -88,17 +89,113 @@ pub enum Message {
     ProbeSocketAck(nt::LightWeightConnectionId),
     CreateNodeId(nt::LightWeightConnectionId, NodeId),
     AckNodeId(nt::LightWeightConnectionId, NodeId),
+    Bytes(nt::LightWeightConnectionId, Bytes),
+
+    UserMessage(M),
+}
 
-    GetBlockHeaders(nt::LightWeightConnectionId, GetBlockHeaders),
-    BlockHeaders(nt::LightWeightConnectionId, Response<BlockHeaders, String>),
-    GetBlocks(nt::LightWeightConnectionId, GetBlocks),
-    Block(nt::LightWeightConnectionId, Response<Block, String>),
-    SendTransaction(nt::LightWeightConnectionId, TxAux),
+pub enum ChainMessage<B, T>
+where
+    B: server::block::BlockService,
+    T: server::transaction::TransactionService,
+{
+    GetBlockHeaders(nt::LightWeightConnectionId, GetBlockHeaders<B::BlockId>),
+    BlockHeaders(
+        nt::LightWeightConnectionId,
+        Response<BlockHeaders<B::Header>, String>,
+    ),
+    GetBlocks(nt::LightWeightConnectionId, GetBlocks<B::BlockId>),
+    Block(nt::LightWeightConnectionId, Response<B::Block, String>),
+    SendTransaction(nt::LightWeightConnectionId, T::TransactionId),
     TransactionReceived(nt::LightWeightConnectionId, Response<bool, String>),
     Subscribe(nt::LightWeightConnectionId, KeepAlive),
-    Bytes(nt::LightWeightConnectionId, Bytes),
 }
-impl Message {
+
+pub trait AsNtEvent
```

Comment:
The naming of both `AsNtEvent` and `to_nt` is confusing with regard to the established naming conventions of conversion traits. `to_nt` should be `into_nt`,
and the trait should be not named with `As*` (implying a by-reference conversion).

Even more Rustic would be to turn the `to_nt` half into `impl From<Message> for nt::Event` and leave `try_from_nt` in a `TryFrom`-like trait (but only because the standard `TryFrom` is unstable).